File: sudoku.rb
class Sudoku
  def initialize(file)
    puzzle = []
    File.open(file) do |file|
      file.each_line do |line|
        row = line.scan(/./).map { |tile| tile.to_i }
        puzzle << row
      end
    end

    puts "Puzzle:"
    show puzzle

    solution = solve puzzle
    if solution.nil?
      puts "Could not find solution"
    else
      puts "Solved:"
      show solution
    end
  end

  def solve(prev_board)
    board, empty_tile = fill(prev_board)
    return board if board.nil? || solved?(board)

    row_i, col_i, possible_values = empty_tile
    possible_values.each do |value|
      copy = clone board
      copy[row_i][col_i] = value
      solved = solve(copy)
      next if solved.nil?
      return solved
    end
    nil
  end

  def fill(board, tile_set = true)
    while tile_set
      empty_tile, tile_set = nil, false

      for_each_empty_tile(board) do |row_i, col_i|
        possible_values = getValues(board, row_i, col_i)
        return nil if possible_values.empty?

        if possible_values.length == 1
          board[row_i][col_i] = possible_values.first
          tile_set = true
        else
          empty_tile = [row_i, col_i, possible_values]
        end
      end
    end
    return board, empty_tile
  end

  def getValues(board, row_i, col_i)
    x = (row_i / 3) * 3
    y = (col_i / 3) * 3
    square = board[x..x+2].flat_map { |row| row[y..y+2] }

    [*(1..9)] - (board[row_i] | board.transpose[col_i] | square)
  end

  def for_each_empty_tile(board)
    board.each_with_index do |row, row_i|
      row.each_with_index do |tile, col_i|
        yield row_i, col_i  if tile.zero?
      end
    end
  end

  def clone(board)
    board.inject([]) { |copy_rows, self_row| copy_rows << self_row.dup }
  end

  def solved?(board)
    board.flatten.none? { |tile| tile.zero? }
  end

  def show(board)
    board.each_with_index do |row, row_i|
      print "+-------"*3 +"+\n" if row_i %3 == 0
      row.each_with_index do |tile, col_i|
        print "| " if col_i % 3 == 0
        if tile.zero? then print "_ " else print "#{tile} " end
      end
      print "|\n"
    end
    print print "+-------"*3 +"+\n"
  end
end

Sudoku.new("puzzles/puzzle2.txt")

File: sudoku.scala
import scala.io._

class Sudoku(val puzzle:String){
  val dim = List.range(0, 9)
  var problem:List[List[Char]] = List()

  Source.fromFile(puzzle).getLines.foreach (line => problem ::= line.toList)

  println("board loaded:")
  show(problem)

  var solution = solve(problem)
  if (isSolved(solution)) {
    println("solved:")
    show(solution)
  } else {
    println("Could not find solution.")
  }

  def solve(p:List[List[Char]]):List[List[Char]] = {
    var (board, emptyTile) = fill(p)
    if (board == null || isSolved(board)) return board

    val (rIndex, cIndex, possibleValues) = emptyTile
    possibleValues.foreach{ (value) =>
      val row = board(rIndex)
      val guess = board.updated(rIndex, row.updated(cIndex, value))
      val solved = solve(guess)
      if (solved != null) return solved
    }
    return null
  }

  def fill(p:List[List[Char]]):(List[List[Char]], (Int, Int, List[Char])) = {
    var board = p
    var tileSet = true
    var emptyTile = (0,0,List[Char]())
    while(tileSet) {
      tileSet = false
      (board, dim).zipped.foreach{ (row, rIndex) =>
        (row, dim).zipped.foreach { (tile, cIndex) =>
          if(tile == '0') {
            var possibleValues = getValues(board, rIndex, cIndex)
            if (possibleValues.size == 0) return (null, emptyTile)
            if (possibleValues.size == 1) {
              board = board.updated(rIndex, row.updated(cIndex, possibleValues(0)))
              tileSet = true
            } else {
              emptyTile = (rIndex, cIndex, possibleValues)
            }
          }
        }
      }
    }
    return (board, emptyTile)
  }

  def getValues(board:List[List[Char]], rIndex:Int, cIndex:Int):List[Char] = {
    val x = (rIndex / 3) * 3
    val y = (cIndex / 3) * 3
    var square:List[Char] = List()
    for (i <- x until x+3) {
      for (j <- y until y+3) {
        square ::= board(i)(j)
      }
    }
    val cols = board.transpose
    return List.range(1, 10).map{v => (v+'0').toChar}.filterNot{v => (board(rIndex) union cols(cIndex) union square).contains(v) }
  }

  def isSolved(board:List[List[Char]]):Boolean = {
    return board != null && !board.flatten.contains('0')
  }

  def show(board:List[List[Char]]) {
    println("+-------+-------+-------+")
    (board, dim).zipped.foreach{ (row, rIndex) =>
      print("| ")
      (row, dim).zipped.foreach { (tile, cIndex) =>
        print(if(tile == '0') "_ " else tile + " ")
        if(cIndex % 3 == 2) print("| ")
      }
      println
      if(rIndex % 3 == 2) println("+-------+-------+-------+")
    }
  }
}

new Sudoku("puzzles/puzzle2.txt");